tb_plugin/torch_tb_profiler/profiler/overall_parser.py

```diff
@@ -177,22 +177,99 @@ def __init__(self):
+                # This runtime is inside this step span. Scan its device_nodes.
+                rt = runtime_node_list[i_runtime]
+                if rt.device_nodes is not None:
+                    for device_node in rt.device_nodes:
+                        step_device_min_ts = min(device_node.start_time, step_device_min_ts)
+                        step_device_max_ts = max(device_node.end_time, step_device_max_ts)
+                        matched_device_nodes.add(device_node)
+                i_runtime += 1
+            elif runtime_node_list[i_runtime].start_time < step_host_end_time:
+                # and runtime_node_list[i_runtime].end_time > step_host_end_time
+                # This runtime intersects with this step span. Skip this runtime.
+                i_runtime += 1
+            else:
+                # runtime_node_list[i_runtime].start_time >= step_host_end_time
+                # This runtime starts after this step's end. Record and move forward this step.
+                steps_device[i_step] = (step_device_min_ts, step_device_max_ts)
+                i_step += 1
+                step_device_min_ts = sys.maxsize
+                step_device_max_ts = -sys.maxsize - 1
+        if i_step < len(self.steps):
+            steps_device[i_step] = (step_device_min_ts, step_device_max_ts)
+            # If there are steps during [i_step+1, len(self.steps) which has no runtime,
+            # then their steps_device will keep as initial value "(sys.maxsize, -sys.maxsize - 1)".
+        # Change step time to device side on the condition that any step have device time.
+        is_use_gpu = False
+        for steps_device_item in steps_device:
+            if steps_device_item[0] != sys.maxsize:
+                is_use_gpu = True
+                break
+        if is_use_gpu:
+            prev_step_end_time = self.steps[0][0]
+            if steps_device[0][0] != sys.maxsize:  # When step 0 has device event.
+                for device_node in device_node_list:
+                    if device_node not in matched_device_nodes:
+                        # Now this device_node is not launched inside any step span.
+                        if device_node.end_time < steps_device[0][0]:
+                            prev_step_end_time = max(prev_step_end_time, device_node.end_time)
+            for i_step in range(len(self.steps)):
+                step_start_time = max(prev_step_end_time, self.steps[i_step][0])
+                step_end_time = self.steps[i_step][1]
+                if steps_device[i_step][0] == sys.maxsize:  # When step i_step has no device event.
+                    # Assign to step_start_time when kernel is behind host step end.
+                    step_end_time = max(step_end_time, step_start_time)
+                else:
+                    step_end_time = max(step_end_time, steps_device[i_step][1])
+                    if step_end_time < step_start_time:
```

Comment:
under what conditions we will hit the start > end?

Reply:
This is tricky, I'm not sure whether it will really happens, just add a guard here. When the step "i_step" launches kernels in a different stream than "i_step-1", this may occur.
![image](https://user-images.githubusercontent.com/10429114/111433352-2646ec80-8739-11eb-8f61-3b0641778a39.png)